pennylane/qnode.py

```diff
@@ -605,8 +605,25 @@ def __call__(self, *args, **kwargs):
             # construct the tape
             self.construct(args, kwargs)
 
-        # execute the tape
-        res = self.qtape.execute(device=self.device)
+        # Execute the tape.
+        # If the observable contains a Hamiltonian and the device does not
+        # support Hamiltonians, split tape into multiple tapes and execute
+        # sequentially. In future, this logic should be moved to the device
+        # to allow for more efficient batch execution.
+        supports_hamiltonian = self.device.supports_observable("Hamiltonian")
+        hamiltonian_in_obs = "Hamiltonian" in [obs.name for obs in self.qtape.observables]
+        if hamiltonian_in_obs and (self.device.shots is not None or not supports_hamiltonian):
+            try:
+                tapes, fn = qml.transforms.hamiltonian_expand(self.qtape, group=False)

```

Comment:
How come grouping is turned off? Would it make sense to allow grouping temporarily, and then turn it off later once grouping is moved into the `Hamiltonian` class?

Reply:
Well, at the moment `ExpvalCost` has `grouping` (called "optimize" or so) turned off by default. And I'd rather activate it once it is moved and tested!

Reply:
Got it, that makes sense

Reply:
@mariaschuld Just curious if you have tested the current PR for supporting batched-shots. There seems to be a bug in ExpvalCost when `optimize=True` is used with, for example, `shot_batch = [(100, 10)]`.

Reply:
Ah that's a good point... we should try and get into the habit of always testing this combination, I admit this is something I forget about!

Reply:
Yes, although the entire idea of setting a shot distribution in the device and then passing a circuit that may be split into a variable number of qnodes is just not working! I hope when grouping and shot distribution are attributes of the Hamiltonian the UI will be intuitive! 

Fixing this in `ExpvalCost` is a major headache with lots of open design questions.